elephant/conversion.py

```diff
@@ -152,3 +155,704 @@ def binarize(spiketrain, sampling_rate=None, t_start=None, t_stop=None,
+        self.matrix_rows = len(spiketrains)
+        # Boolean, indicates if matrix shall stored
+        self.store_mat_u = store_mat
+        # Empty matrix for storage, unclipped matrix
+        self.mat_u = None
+        # Check all parameter, set also missing values
+        self.__calc_start_stop(spiketrains)
+        self.__check_init_params(binsize, num_bins, self.t_start, self.t_stop)
+        self.__check_consistency(spiketrains, self.binsize, self.num_bins,
+                                 self.t_start, self.t_stop)
+        # Variables to store the sparse matrices
+        self._sparse_mat_u = None
+        self._sparse_mat_c = None
+        # Now create unclipped version of sparse matrix
+        self.__convert_to_binned(spiketrains)
+
+    # =========================================================================
+    # There are four cases the given parameters must fulfill
+    # Each parameter must be a combination of following order or it will raise
+    # a value error:
+    # t_start, num_bins, binsize
+    # t_start, num_bins, t_stop
+    # t_start, bin_size, t_stop
+    # t_stop, num_bins, binsize
+    # ==========================================================================
+
+    def __check_init_params(self, binsize, num_bins, t_start, t_stop):
+        """
+        Checks given parameter.
+        Calculates also missing parameter.
+
+        Parameters
+        ----------
+        binsize : quantity.Quantity
+            Size of Bins
+        num_bins : int
+            Number of Bins
+        t_start: quantity.Quantity
+            Start time of the spike
+        t_stop: quantity.Quantity
+            Stop time of the spike
+
+        Raises
+        ------
+        ValueError :
+            If all parameters are `None`, a ValueError is raised.
+        TypeError:
+            If type of :attr:`num_bins` is not an Integer.
+        """
+        # Raise error if no argument is given
+        if binsize is None and t_start is None and t_stop is None \
+                and num_bins is None:
+            raise ValueError(
+                "No arguments given. Please enter at least three arguments")
+        # Check if num_bins is an integer (special case)
+        if num_bins is not None:
+            if type(num_bins) is not int:
+                raise TypeError("num_bins is not an integer!")
+        # Check if all parameters can be calculated, otherwise raise ValueError
+        if t_start is None:
+            self.t_start = calc_tstart(num_bins, binsize, t_stop)
+        elif t_stop is None:
+            self.t_stop = calc_tstop(num_bins, binsize, t_start)
+        elif num_bins is None:
+            self.num_bins = calc_num_bins(binsize, t_start, t_stop)
+            if self.matrix_columns is None:
+                self.matrix_columns = self.num_bins
+        elif binsize is None:
+            self.binsize = calc_binsize(num_bins, t_start, t_stop)
+
+    def __calc_start_stop(self, spiketrains):
```

Comment:
Since this has no class-breaking side effects, I would make this (and similar functions) single underscore (`_calc_start_stop` not `__calc_start_stop`) to disable name mangling.